accepted/future-releases/constructor-tearoffs/test-plan.md

```diff
@@ -0,0 +1,168 @@
+# Dart Constructor Tear-Offs Test Plan
+
+The Dart constructor tear-off feature contains multiple independent language changes. Each should be tested, as should their combination with each other and with other relevant language features.
+
+The features can be summarized as:
+
+* Named constructor tear-offs
+  * Expressions of the for <code>*C*.*name*</code>. or <code>*C*\<*typeArgs*>.*name*</code>, with type arguments potentially being inferred, evaluate to functions.
+  * Preserves identity as specified.
+* Unnamed constructor syntax alternative
+  * Everywhere you can currently reference an unnamed constructor, you can also refer to it as <code>*C*.new</code>.
+  * The unnamed and `new`-named constructors are two ways to refer to the *same* thing. (Can't declare both).
+* Unnamed constructor tear-offs
+  * Expressions of the for <code>*C*.new</code>. or <code>*C*\<*typeArgs*>.new</code>, with type arguments potentially being inferred, evaluate to functions.
+  * Preserves identity as specified.
+* Explicit instantiated function tear-off
+  * Expressions of the form <code>*f*\<*typeArgs*></code>, not followed by an argument list and where *f* denotes a generic function declaration or generic instance method, now does instantiated tear-off.
+* Explicit type literal instantiation
+  * Expressions of the form <code>*T*\<*typeArgs*></code>, not followed by an argument list and where *T* denotes a type declaration (class, mixin or type alias), now evaluates to a `Type` object for the instantiated type.
+
+Each feature should be tested, both for correct and incorrect usage. Further, we should check that any new syntax is not allowed in places where it isn't defined.
+
+## Named constructor tear-off
+
+### Correct usage
+
+The possible constructors tear-offs
+
+* Check that constructors of non-generic classes can be torn off.
+* Check that constructors of instantiated generic classes can be torn off.
+* Check that constructors of uninstantiated generic classes can be torn off, and that type inference infers the correct type arguments.
+* Check that the resulting function has the correct static type and runtime type.
+
+The possible constructor types (all combinations except "const factory non-redirecting" exist):
+
+* Constant or not
+* Factory or not
+* Redirecting or not
+
+The possible sources of tear-offs:
+
+* From a class declaring the constructor.
+* From a mixin-application which introduces a forwarding constructor.
+* Any of those through an alias which is:
+  * Not generic.
+  * A proper rename (static type may not match runtime type, static type is based on alias type argument bounds, runtime type is based on class type argument bounds).
```

Comment:
Specifically need to check that tearoffs from proper renames of generics aren't treated as instantiate to bounds, but instead are subject to normal instance creation inference at a call site (e.g. `C<int> x = TypeDef.name(3)` should use the context type to fill in the missing type argument to `TypeDef`).

Reply:
Also should ensure that bounds are checked.